pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/impl/v1/VarByteSingleValueReader.java

```diff
@@ -0,0 +1,166 @@
+import com.linkedin.pinot.core.segment.memory.PinotDataBuffer;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+
+
+/**
+ * Reader class for data written out by {@link SingleColumnSingleValueWriter}.
+ */
+public class VarByteSingleValueReader extends BaseSingleColumnSingleValueReader<VarByteReaderContext> {
+  private static final int INT_SIZE = Integer.SIZE / Byte.SIZE;
+  private static final Charset UTF_8 = Charset.forName("UTF-8");
+
+  final PinotDataBuffer _dataBuffer;
+  final PinotDataBuffer _header;
+  private final ChunkUncompressor _chunkUncompressor;
+  private final int _maxChunkSize;
+
+  private final int _numDocsPerChunk;
+  private final int _numChunks;
+  private final int _lengthOfLongestEntry;
+
+  // Thread local (reusable) byte[] to read bytes from data file.
+  private final ThreadLocal<byte[]> _reusableBytes = new ThreadLocal<byte[]>() {
+    @Override
+    protected byte[] initialValue() {
+      return null;
+    }
+  };
+
+  /**
+   * Constructor for the class.
+   *
+   * @param pinotDataBuffer Data buffer to read from
+   * @param uncompressor Chunk uncompressor
+   * @throws IOException
+   */
+  public VarByteSingleValueReader(PinotDataBuffer pinotDataBuffer, ChunkUncompressor uncompressor)
+      throws IOException {
+    _chunkUncompressor = uncompressor;
+    _dataBuffer = pinotDataBuffer;
+
+    int headerOffset = INT_SIZE; // First entry is the version, which is unused currently.
+    _numChunks = _dataBuffer.getInt(headerOffset);
+    headerOffset += INT_SIZE;
+
+    _numDocsPerChunk = _dataBuffer.getInt(headerOffset);
+    headerOffset += INT_SIZE;
+
+    _lengthOfLongestEntry = _dataBuffer.getInt(headerOffset);
+    headerOffset += INT_SIZE;
+
+    int chunkHeaderSize = _numDocsPerChunk * INT_SIZE;
+    _maxChunkSize = chunkHeaderSize + (_lengthOfLongestEntry * _numDocsPerChunk);
+
+    // Slice out the header from the data buffer.
+    int headerLength = _numChunks * INT_SIZE;
+    _header = _dataBuffer.view(headerOffset, headerOffset + headerLength);
+  }
+
+  @Override
+  public String getString(int row, VarByteReaderContext context)
+      throws IOException {
+
+    int chunkRowId = row % _numDocsPerChunk;
+    ByteBuffer chunkBuffer = getChunkForRow(row, context);
+
+    int rowOffset = chunkBuffer.getInt(chunkRowId * INT_SIZE);
+    int nextRowOffset = chunkBuffer.getInt((chunkRowId + 1) * INT_SIZE);
+
+    // For incomplete chunks, the next string's offset will be 0 as row offset for absent rows are 0.
+    if ((chunkRowId == (_numDocsPerChunk - 1)) || (nextRowOffset == 0)) {
+      nextRowOffset = chunkBuffer.limit();
+    }
+
+    int length = nextRowOffset - rowOffset;
+    chunkBuffer.position(rowOffset);
+
+    // Lazy allocation of thread local memory.
+    byte[] bytes = _reusableBytes.get();
+    if (bytes == null) {
+      bytes = new byte[_lengthOfLongestEntry];
+      _reusableBytes.set(bytes);
+    }
+
+    chunkBuffer.get(bytes, 0, length);
```

Comment:
(Critical) I believe this way to get bytes is not thread-safe as you reuse the same chunk here.